app/src/main/java/com/nervousfish/nervousfish/modules/database/GsonDatabaseAdapterLoader.java

```diff
@@ -0,0 +1,134 @@
+    private static final Type TYPE_DATABASE = new TypeToken<Database>() {
+    }.getType();
+    private static final String BAD_PADDING = "Database has bad padding";
+    private static final String DATABASE_WRONG_SIZE = "Database has the wrong size";
+
+    private final IFileSystem fileSystem;
+    private final IEncryptor encryptor;
+    private final String databasePath;
+    private final String passwordPath;
+
+
+    /**
+     * Helper class for GsonDatabaseAdapter.
+     *
+     * @param serviceLocator Can be used to get access to other modules
+     */
+    GsonDatabaseAdapterLoader(final IServiceLocator serviceLocator) {
+        this.fileSystem = serviceLocator.getFileSystem();
+        this.encryptor = serviceLocator.getEncryptor();
+        this.databasePath = serviceLocator.getConstants().getDatabasePath();
+        this.passwordPath = serviceLocator.getConstants().getPasswordPath();
+    }
+
+    /**
+     * Reads the database to a string.
+     *
+     * @return A String containing the json database
+     */
+    private String readDatabaseToString() throws IOException {
+        final StringBuilder databaseFileStringBuilder = new StringBuilder();
+
+        // Get the database from the database file
+        try (BufferedReader databaseReader = (BufferedReader) this.fileSystem.getReader(databasePath)) {
+
+            while (true) {
+                final String line = databaseReader.readLine();
+                if (line == null) {
+                    break;
+                }
+                databaseFileStringBuilder.append(line);
+            }
+        }
+        return databaseFileStringBuilder.toString();
+    }
+
+    /**
+     * Checks the password, throws IOException if it's wrong.
+     *
+     * @param password -   The password to check.
+     * @throws IOException Thrown when password is wrong.
+     */
+    private void checkPassword(final String password) throws IOException {
+
+        final BufferedReader passReader = (BufferedReader) this.fileSystem.getReader(passwordPath);
+        final StringBuffer passwordFileStringBuffer = new StringBuffer();
+        while (true) {
+            final String line = passReader.readLine();
+            if (line == null) {
+                break;
+            }
+            passwordFileStringBuffer.append(line);
+        }
+        final String encryptedPassword = passwordFileStringBuffer.toString();
+        if (!encryptedPassword.equals(encryptor.hashWithoutSalt(password))) {
+            throw new IOException("Password is wrong " + encryptedPassword + " " + encryptor.hashWithoutSalt(password));
+        }
+        passReader.close();
+    }
+
+    /**
+     * Loads the Secretkey from a given password.
+     *
+     * @param password THe password to load a key from
+     * @return The secretKey made from the password\
+     */
+    public SecretKey loadKey(final String password) throws IOException, InvalidKeySpecException {
```

Comment:
I guess you should be able to obtain the secret key from an `IEncryptor` instance whenever you need it